wallet/src/account/output_cache/mod.rs

```diff
@@ -1401,89 +1437,147 @@ impl OutputCache {
-            all_abandoned.push(*outpoint_source_id.get_tx_id().expect("must be a transaction"));
+        while let Some(outpoint_source_id) = to_update.pop_first() {
+            all_txs.push(*outpoint_source_id.get_tx_id().expect("must be a transaction"));
 
             if let Some(descendants) = self.unconfirmed_descendants.remove(&outpoint_source_id) {
-                to_abandon.extend(descendants.into_iter())
+                to_update.extend(descendants.into_iter())
+            }
+        }
+
+        all_txs
+    }
+
+    // After tx is removed as a result of reorg, abandoning or marking as conflicted
+    // its effect on OutputCache's data fields should be rolled back
+    fn rollback_tx_data(&mut self, tx: &WalletTx) -> WalletResult<()> {
+        let tx_id = tx.id();
+
+        // Iterate in reverse to handle situations where an account is modified twice in the same tx
+        for input in tx.inputs().iter().rev() {
+            match input {
+                TxInput::Utxo(outpoint) => {
+                    self.consumed.remove(outpoint);
+                }
+                TxInput::Account(outpoint) => match outpoint.account() {
+                    AccountSpending::DelegationBalance(delegation_id, _) => {
+                        if let Some(data) = self.delegations.get_mut(delegation_id) {
+                            data.last_nonce = outpoint.nonce().decrement();
+                            data.last_parent = find_parent(&self.unconfirmed_descendants, &tx_id);
+                        }
+                    }
+                },
+                TxInput::AccountCommand(nonce, op) => match op {
+                    AccountCommand::MintTokens(token_id, _)
+                    | AccountCommand::UnmintTokens(token_id)
+                    | AccountCommand::LockTokenSupply(token_id)
+                    | AccountCommand::FreezeToken(token_id, _)
+                    | AccountCommand::UnfreezeToken(token_id)
+                    | AccountCommand::ChangeTokenMetadataUri(token_id, _)
+                    | AccountCommand::ChangeTokenAuthority(token_id, _) => {
+                        if let Some(data) = self.token_issuance.get_mut(token_id) {
+                            data.last_nonce = nonce.decrement();
+                            data.last_parent = find_parent(&self.unconfirmed_descendants, &tx_id);
+                            data.unconfirmed_txs.remove(&tx_id);
+                        }
+                    }
+                    AccountCommand::ConcludeOrder(order_id)
+                    | AccountCommand::FillOrder(order_id, _, _) => {
+                        if let Some(data) = self.orders.get_mut(order_id) {
+                            data.last_nonce = nonce.decrement();
+                            data.last_parent = find_parent(&self.unconfirmed_descendants, &tx_id);
+                        }
+                    }
+                },
+            }
+        }
+
+        for output in tx.outputs() {
```

Comment:
Could be nitpicking, but I'd suggest rolling back stuff in the exact opposit order of how it was added. I.e. first roll back outputs (in the reverse order too) and then inputs.
Currently it doesn't matter of course, but maybe our chain will evolve in the future and make the order matter.

Reply:
I changed the order to rev(), but processing outputs before inputs break logic that searches for utxo parent, so I'd leave it this way